File: nodejs/day13/index1.js
const { readLines, parseInt } = require("../lib");

const parse = line => line.split(/\s*:\s*/).map(parseInt);

let maxLayer = 0;

const data = readLines("./input")
  .map(parse)
  .reduce((acc, [l, d]) => {
    acc[l] = d;
    maxLayer = l;
    return acc;
  }, []);

const getPos = (d, t) => {
  const m = 2 * d - 2;
  const i = t % m;
  if (i >= d) {
    return m - i;
  }
  return i;
};

let s = 0;

for (let i = 0; i <= maxLayer; i++) {
  if (data[i] == null) {
    continue;
  }
  if (getPos(data[i], i) === 0) {
    s += i * data[i];
  }
}

console.log(s);

File: clojure/day13/index1.clj
(require '[clojure.string :as s])

(def parse-int #(Integer/parseInt %))
(defn split [p s] (s/split s p))

(defn parse-line [line] (->> line (split #"\s*:\s*") (map parse-int)))

(def data (->>
  "./input"
  slurp
  s/trim-newline
  s/split-lines
  (map parse-line)
  (reduce (fn [acc [l d]] (assoc acc l d)) {})))

(def max-layer (apply max (keys data)))

(defn get-pos [d t] (let [
  m (- (* 2 d) 2)
  i (mod t m)
] (if (>= i d) (- m i) i)))

(def res (loop [
  i 0
  s 0
] (if (> i max-layer) s
    (let [
      d (get data i)
      s' (if (and d (== 0 (get-pos d i)))
          (+ s (* i d))
          s)
    ] (recur (inc i) s')))))

(println res)